res/res_pjsip_device_features.c

```diff
@@ -0,0 +1,1005 @@
+		res = -1;
+	}
+
+cleanup:
+	ast_xml_close(xmldoc);
+	return res;
+}
+
+static void subscription_shutdown(struct ast_sip_subscription *sub);
+static int new_subscribe(struct ast_sip_endpoint *endpoint, const char *resource);
+static int refresh_subscribe(struct ast_sip_subscription *sub, pjsip_rx_data *rdata);
+static int subscription_established(struct ast_sip_subscription *sub);
+static void *get_notify_data(struct ast_sip_subscription *sub);
+static int get_resource_display_name(struct ast_sip_endpoint *endpoint, const char *resource, char *display_name, int display_name_size);
+static void to_ami(struct ast_sip_subscription *sub, struct ast_str **buf);
+
+struct ast_sip_notifier feature_notifier = {
+	.default_accept = FEATURE_SYNC_MIME_TYPE,
+	.new_subscribe = new_subscribe,
+	.refresh_subscribe = refresh_subscribe,
+	.subscription_established = subscription_established,
+	.get_notify_data = get_notify_data,
+	.get_resource_display_name = get_resource_display_name,
+};
+
+struct ast_sip_subscription_handler feature_handler = {
+	.event_name = FEATURE_SYNC_EVENT,
+	.body_type = AST_SIP_DEVICE_FEATURE_SYNC_DATA,
+	.accept = { FEATURE_SYNC_MIME_TYPE, },
+	.subscription_shutdown = subscription_shutdown,
+	.to_ami = to_ami,
+	.notifier = &feature_notifier,
+};
+
+static void feature_state_subscription_destructor(void *obj)
+{
+	struct feature_state_subscription *sub = obj;
+	ast_sip_subscription_destroy(sub->sip_sub);
+	ast_taskprocessor_unreference(sub->serializer);
+}
+
+static inline void update_everything(struct feature_state_subscription *feature_state_sub, struct ast_sip_device_feature_sync_data *sync_data)
+{
+	if (feature_state_sub->have_dnd) {
+		sync_data->update_needed_dnd = 1;
+	}
+	if (feature_state_sub->have_callforwardalways) {
+		sync_data->update_needed_fwd_always = 1;
+	}
+	if (feature_state_sub->have_callforwardbusy) {
+		sync_data->update_needed_fwd_noanswer = 1;
+	}
+	if (feature_state_sub->have_callforwardnoanswer) {
+		sync_data->update_needed_fwd_busy = 1;
+	}
+}
+
+static inline void clear_everything(struct ast_sip_device_feature_sync_data *sync_data)
+{
+	sync_data->update_needed_dnd = 0;
+	sync_data->update_needed_fwd_always = 0;
+	sync_data->update_needed_fwd_noanswer = 0;
+	sync_data->update_needed_fwd_busy = 0;
+}
+
+/*!
+ * \internal
+ * \brief Allocates an feature_state_subscription object.
+ *
+ * Creates the underlying SIP subscription for the given request. First makes
+ * sure that there are registered handler and provider objects available.
+ */
+static struct feature_state_subscription *feature_state_subscription_alloc(struct ast_sip_subscription *sip_sub, struct ast_sip_endpoint *endpoint)
+{
+	struct feature_state_subscription *feature_state_sub;
+	struct subscription_item *subitem;
+	char tmp[16];
+	const char *endpoint_name = ast_sorcery_object_get_id(endpoint);
+
+	subitem = ast_calloc(1, sizeof(*subitem));
+	if (!subitem) {
+		return NULL;
+	}
+	subitem->endpoint = ast_strdup(endpoint_name);
```

Comment:
This doesn't need to be a separate allocation, it can be allocated as part of subitem